tests/phpunit/CRM/Civicase/Hook/BuildForm/TokenTreeTest.php

```diff
@@ -0,0 +1,329 @@
+            ],
+          ],
+        ],
+        [
+          'text' => 'Current User',
+          'children' => [
+            [
+              'id' => '{current_user.contact_city}',
+              'text' => 'Current User City',
+            ],
+            [
+              'id' => '{current_user.contact_custom_' . $this->contactCustomField['id'] . '}',
+              'text' => 'Current User ' . $this->contactCustomField['name'],
+            ],
+          ],
+        ],
+        [
+          'text' => 'Case',
+          'children' => [
+            [
+              'id' => '{case.id}',
+              'text' => 'Case Id',
+            ],
+          ],
+        ],
+        [
+          'text' => 'Contact',
+          'children' => [
+            [
+              'id' => '{contact.addressee_id}',
+              'text' => 'Addressee ID',
+            ],
+          ],
+        ],
+        [
+          'text' => $this->contactCustomField['name'],
+          'children' => [
+            [
+              'id' => '{contact.custom_' . $this->contactCustomField['id'] . '}',
+              'text' => $this->contactCustomField['name'],
+            ],
+          ],
+        ],
+        [
+          'text' => $this->caseCustomField['name'],
+          'children' => [
+            [
+              'id' => '{case.custom_' . $this->caseCustomField['id'] . '}',
+              'text' => $this->caseCustomField['name'],
+            ],
+          ],
+        ],
+    ];
+  }
+
+  /**
+   * Format new token tree.
+   *
+   * @param array $tokens
+   *   List of tokens.
+   *
+   * @return array
+   *   Formatted tokens.
+   */
+  private function format(array $tokens) {
+    $formattedTokens = [];
+    foreach ($tokens as $token) {
+      $formattedTokens[$token['text']] = $token;
+    }
+
+    return $formattedTokens;
+  }
+
+  /**
+   * Verify current user tokens.
+   *
+   * @param array $newTokenTree
+   *   List of tokens.
+   */
+  private function verifyCurrentUserTokens(array $newTokenTree) {
+    $this->assertNotEmpty($newTokenTree['Current User']);
+    $this->assertEquals(
+      '{current_user.contact_city}',
+      $newTokenTree['Current User']['children'][0]['children'][0]['id']
+    );
+    $this->assertEquals(
+      'Current User City',
+      $newTokenTree['Current User']['children'][0]['children'][0]['text']
+    );
+    $this->assertEquals(
+      '{current_user.contact_custom_' . $this->contactCustomField['id'] . '}',
+      $newTokenTree['Current User']['children'][1]['children'][0]['children'][0]['id']
+    );
+    $this->assertEquals(
+      'Current User ' . $this->contactCustomField['name'],
+      $newTokenTree['Current User']['children'][1]['children'][0]['children'][0]['text']
+    );
+  }
+
+  /**
+   * Verify case role tokens.
+   *
+   * @param array $newTokenTree
+   *   List of tokens.
+   */
+  private function verifyCaseRoleTokens(array $newTokenTree) {
```

Comment:
The assertions here seem pretty long. Also we are relying pretty much on the names of pre-installed relationship types. What of the name changes, the tests would probably break.
I would suggest we do an API call to fetch two of the pre-installed relationships (Sort ASC or whichever is appropriate) and then loop through those to verify the property for each of the relationship types like id, contact_sub_type and so on so that it is generic and we can even test for more relationship types without adding to the line of code, we only increase the number fetched by the API.

Reply:
well we dont rely on the names of the relationship from database I am creating  the relationship and attaching it to the form in gettokens function so it wont break if names changes in the db as it does not depend on db